versions/1.1/SPEC.md

```diff
@@ -0,0 +1,5846 @@
+
+```sh
+sh script.sh /jobs/564757/sample_quality_scores.tsv
+```
+
+Where `/jobs/564757/sample_quality_scores.tsv` would contain:
+
+```txt
+sample1\t98
+sample2\t95
+sample3\t75
+```
+
+#### Map serialization using write_json()
+
+The `Map` value can also be serialized as a JSON `object` in a file, and the parameter on the command line is given the path to that file, using the `write_json()` function:
+
+```wdl
+task test {
+  input {
+    Map[String, Float] sample_quality_scores
+  }
+
+  command {
+    sh script.sh ${write_json(sample_quality_scores)}
+  }
+  
+  runtime {
+    container: "my_image:latest"
+  }
+}
+```
+
+If `sample_quality_scores` were a `Map` with these members:
+
+|Key    |Value |
+|-------|------|
+|sample1|98    |
+|sample2|95    |
+|sample3|75    |
+
+Then, the resulting command line might look like:
+
+```sh
+sh script.sh /jobs/564757/sample_quality_scores.json
+```
+
+Where `/jobs/564757/sample_quality_scores.json` would contain:
+
+```json
+{
+  "sample1": 98,
+  "sample2": 95,
+  "sample3": 75
+}
+```
+
+#### Map deserialization using read_map()
+
+`read_map()` will return a `Map[String, String]` where the keys are the first column in the TSV input file and the corresponding values are the second column.
+
+This return value can be coerced to other `Map` types. For example:
+
+```wdl
+task test {
+  command <<<
+    python <<CODE
+    for i in range(3):
+      print("key_{idx}\t{idx}".format(idx=i))
+    CODE
+  >>>
+
+  output {
+    Map[String, Int] my_ints = read_map(stdout())
+  }
+  
+  runtime {
+    container: "my_image:latest"
+  }
+}
+```
+
+`my_ints` will be a `Map[String, Int]` with members:
+
+|Key  |Value |
+|-----|------|
+|key_0|0     |
+|key_1|1     |
+|key_2|2     |
+
+#### Map deserialization using read_json()
+
+`read_json()` will return whatever data type resides in the JSON file. If the file contains a JSON object, it is deserialized as an `Object` value, which can be coerced to a `Map[String, X]` so long as all values are coercible to `X`.
+
+```wdl
+task test {
+  command <<<
+    echo '{"foo":"bar"}'
+  >>>
+
+  output {
+    Map[String, String] my_map = read_json(stdout())
+  }
+  
+  runtime {
+    container: "my_image:latest"
+  }
+}
+```
+
+`my_map` will be a `Map[String, String] with members:
+
+|Key |Value |
+|----|------|
+|foo |bar   |
+
+### Pair
+
+There are no functions to directly serialize and deserialize `Pair`s. Instead, the left and right values should be passed to the command independently.
+
+In this script, `myscript` writes two files, `int.txt` containing an `Int`, and `str.txt` containing a `String`.
+
+```wdl
+task pair_test {
+  input {
+    Pair[Int, String] p
+  }
+
+  command <<<
+  ./myscript -i ~{p.left} -s ~{p.right}
+  >>>
+
+  output {
+    Pair[Int, String] out = (read_int("int.txt"), read_string("str.txt")) 
+  }
+}
+```
+
+Alternatively, a `struct` or `Object` can be created with `left` and `right` members and then serialized by the appropriate function.
```

Comment:
Perhaps remove this example using Object and make it use struct instead?

Reply:
done